From 85dcbd562df0230638c8a646b724882b53966466 Mon Sep 17 00:00:00 2001
From: Christophe Priouzeau <christophe.priouzeau@linaro.org>
Date: Wed, 6 Apr 2016 15:46:04 +0200
Subject: [PATCH 05/13] STM: streamsplitter : push pending events before
 serialized query

set proxy allocation flag on stream splitter sinkpad so that it
forwards the allocation queries instead of discarding them.
push pending events downstream on active pad before forwarding
a serialized query, otherwise allocation query reaches downstream
before caps event resulting into unanswered allocation query.

Change-Id: I1f7c4d013fbb248b9c9bd443960c35abe9fa13aa
Signed-off-by: Rajesh Sharma <rajesh-dcg.sharma@st.com>
---
 gst/encoding/gststreamsplitter.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/gst/encoding/gststreamsplitter.c b/gst/encoding/gststreamsplitter.c
index 6bf3894..f5eae4d 100644
--- a/gst/encoding/gststreamsplitter.c
+++ b/gst/encoding/gststreamsplitter.c
@@ -382,6 +382,8 @@ static gboolean
 gst_stream_splitter_sink_query (GstPad * pad, GstObject * parent,
     GstQuery * query)
 {
+  GstStreamSplitter *stream_splitter = (GstStreamSplitter *) parent;
+  GstPad *srcpad = NULL;
   gboolean res;
 
   switch (GST_QUERY_TYPE (query)) {
@@ -408,6 +410,17 @@ gst_stream_splitter_sink_query (GstPad * pad, GstObject * parent,
       break;
     }
     default:
+      if (GST_QUERY_IS_SERIALIZED (query)) {
+        STREAMS_LOCK (stream_splitter);
+        if (stream_splitter->current)
+          srcpad = gst_object_ref (stream_splitter->current);
+        STREAMS_UNLOCK (stream_splitter);
+
+        if (G_UNLIKELY (stream_splitter->pending_events) && (srcpad != NULL)) {
+          gst_stream_splitter_push_pending_events (stream_splitter, srcpad);
+          gst_object_unref (srcpad);
+        }
+      }
       res = gst_pad_query_default (pad, parent, query);
       break;
   }
@@ -478,6 +491,7 @@ gst_stream_splitter_init (GstStreamSplitter * stream_splitter)
       gst_stream_splitter_sink_event);
   gst_pad_set_query_function (stream_splitter->sinkpad,
       gst_stream_splitter_sink_query);
+  GST_PAD_SET_PROXY_ALLOCATION (stream_splitter->sinkpad);
   gst_element_add_pad (GST_ELEMENT (stream_splitter), stream_splitter->sinkpad);
 
   g_mutex_init (&stream_splitter->lock);
-- 
2.7.4

